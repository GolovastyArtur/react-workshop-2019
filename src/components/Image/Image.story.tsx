import React from 'react';

import {storiesOf} from '@storybook/react';

import {Image} from './Image';

storiesOf('Image', module)
    .add('default', () => (
        <Image
            src="https://avatars.mds.yandex.net/get-pdb/27625/7c81ce06-f70e-405b-87b4-fe826ba9362e/s375"
        />
    ));
