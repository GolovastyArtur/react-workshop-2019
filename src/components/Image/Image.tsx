import * as React from 'react';

import classnames from 'classnames';

export interface ImageProps {
    src?: string;
    alt?: string;
    ratio?: number;
    className?: string;
}

interface ComponentState {
    hidden: boolean;
}

export class Image extends React.Component<ImageProps, ComponentState> {
    public state: ComponentState = {
        hidden: true,
    };

    private image: React.RefObject<HTMLImageElement> = React.createRef();

    onLoad() {
        let image = this.image.current;

        if (image && this.state.hidden) {
            if (image.naturalHeight > 0) {
                this.setState({hidden: false});
            }
        }
    }

    render() {
        let {src, alt, ratio, className} = this.props;
        let {hidden} = this.state;

        if (!src) {
            return null;
        }

        return (
            <div
                className={classnames('image', className, {
                    'image_hidden': hidden,
                })}
                style={{
                    minHeight: ratio ? `${ratio * 100}vw` : undefined,
                }}
            >
                <img
                    src={src}
                    className="image__image"
                    alt={alt}
                    onLoad={() => this.onLoad()}
                    ref={this.image}
                />
            </div>
        );
    }

}
