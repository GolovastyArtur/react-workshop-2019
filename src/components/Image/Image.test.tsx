import * as React from 'react';
import * as renderer from 'react-test-renderer';

import {Image} from './Image';

describe('Image', function () {
    it('no image', function() {
        let component = renderer.create(<Image />);

        expect(component.toJSON()).toMatchSnapshot();
    });

    it('with image', function() {
        let component = renderer.create(
            <Image
                src="https://avatars.mds.yandex.net/get-pdb/27625/7c81ce06-f70e-405b-87b4-fe826ba9362e/s375"
            />
        );

        expect(component.toJSON()).toMatchSnapshot();
    });
});
