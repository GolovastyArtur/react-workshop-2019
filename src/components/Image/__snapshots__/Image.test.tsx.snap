// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image no image 1`] = `null`;

exports[`Image with image 1`] = `
<div
  className="image image_hidden"
  style={
    Object {
      "minHeight": undefined,
    }
  }
>
  <img
    className="image__image"
    onLoad={[Function]}
    src="https://avatars.mds.yandex.net/get-pdb/27625/7c81ce06-f70e-405b-87b4-fe826ba9362e/s375"
  />
</div>
`;
